### PR TITLE
[HTTP2ConnectionPool] added `HTTP2StateMachine`

### PR DESCRIPTION
**Pull request by dnadoba**
One step closer to support HTTP/2 in the new connection pool. 
`HTTP2StateMaschine ` will be integrated in `HTTPConnectionPool.StateMachine` in a follow up PR.

### CODE REVIEW

File: Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+                // When a connection close is initiated by the connection pool, the connection will
+                // still report its close to the state machine. In those cases we must ignore the
+                // event.
+                return .none
+            }
+            return self.nextActionForFailedConnection(at: index, on: context.eventLoop)
+        }
+
+        mutating func http1ConnectionReleased(_: Connection.ID) -> Action {
+            fatalError("TODO: implement \(#function)")
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift:380]:
```suggestion
            // It is save to bang the http1Connections here. If we get this callback but we don't have 
            // http1 connections something has gone terribly wrong.
            let (index, _) = self.http1Connections!.releaseConnection(connectionID)
            // Any http1 connection that becomes idle should be closed right away after the transition
            // to http2.
            let connection = self.http1Connections.closeConnection(at: index)
            if self.http1Connections!.isEmpty {
                self.http1Connections = nil
            }
            return connection
```

File: Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+        }
+
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift:382]:
```swift
mutating func http1ConnectionClosed(_ connectionID: Connection.ID) -> Action {
    guard let index = self.http1Connections?.failConnection(connectionID)?.0 else {
        return .none
    }
    self.http1Connections!.removeConnection(at: index)
    if self.http1Connections!.isEmpty {
        self.http1Connections = nil
    }
    return .none
}
```

File: Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+            precondition(self.state == .running, "If we are shutting down, we must not have any idle connections")
+
+            return .init(
+                request: .none,
+                connection: .closeConnection(connection, isShutdown: .no)
+            )
+        }
+
+        mutating func http1ConnectionClose(_ connectionID: Connection.ID) -> Action {
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift:372]: I think this is missing a `closeD`

File: Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+                return .init(request: .none, connection: .closeConnection(connection, isShutdown: .no))
+            case .shuttingDown(let unclean):
+                if self.http1Connections == nil && self.connections.isEmpty {
+                    return .init(request: .none, connection: .closeConnection(connection, isShutdown: .yes(unclean: unclean)))
+                } else {
+                    return .init(request: .none, connection: .closeConnection(connection, isShutdown: .no))
+                }
+            case .shutDown:
+                preconditionFailure("If the pool is already shutdown, all connections must have been torn down.")
+            }
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift:418]:
```suggestion
            guard self.http1Connections!.isEmpty else {
                return .init(request: .none, connection: .closeConnection(connection, isShutdown: .no))
            }
            // if there are no more http1Connections, we can remove the struct.
            self.http1Connections = nil
            
            // we must also check, if we are shutting down. Was this maybe out last connection?
            switch state {
            case .running:
                return .init(request: .none, connection: .closeConnection(connection, isShutdown: .no))

            case .shuttingDown(let unclean):
                if self.connections.isEmpty {
                    // if the http2connections are empty as well, there are no more connections. Shutdown completed.
                    return .init(request: .none, connection: .closeConnection(connection, isShutdown: .yes(unclean: unclean)))
                } else {
                    return .init(request: .none, connection: .closeConnection(connection, isShutdown: .no))
                }

            case .shutDown:
                preconditionFailure("If the pool is already shutdown, all connections must have been torn down.")
            }
```

File: Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+RequestQueue.swift [hunk 5]
```diff
@@ -118,3 +133,24 @@ extension HTTPConnectionPool {
...
+    /// Calling this method may invalidate any existing indices for use with this
+    /// `CircularBuffer`.
+    ///
+    /// - Parameter max: The number of elements to remove.
+    ///   `max` must be greater than or equal to zero.
+    /// - Returns: removed elements
+    ///
+    /// - Complexity: O(*k*), where *k* is the number of elements removed.
+    fileprivate mutating func popFirst(max: Int) -> [Element] {
+        precondition(max >= 0, "")
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+RequestQueue.swift:150]: Add a message, or remove the empty string :)

File: Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+                requestAction = .failRequestsAndCancelTimeouts(waitingRequests, HTTPClientError.cancelled)
+            }
+
+            // clean up the connections, we can cleanup now!
+            let cleanupContext = self.connections.shutdown()
+
+            // If there aren't any more connections, everything is shutdown
+            let isShutdown: StateMachine.ConnectionAction.IsShutdown
+            let unclean = !(cleanupContext.cancel.isEmpty && waitingRequests.isEmpty)
+            if self.connections.isEmpty {
```
Comment (fabianfett) [inline Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2StateMachine.swift:437]: We also need to check if we have still running http1 connections here.

File: Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+RequestQueue.swift [hunk 7]
```diff
@@ -118,3 +133,24 @@ extension HTTPConnectionPool {
...
+    ///
+    /// Calling this method may invalidate any existing indices for use with this
+    /// `CircularBuffer`.
+    ///
+    /// - Parameter max: The number of elements to remove.
+    ///   `max` must be greater than or equal to zero.
+    /// - Returns: removed elements
+    ///
+    /// - Complexity: O(*k*), where *k* is the number of elements removed.
+    fileprivate mutating func popFirst(max: Int) -> [Element] {
```
[the remark this answers is not in the document]
Comment (fabianfett) [reply Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+RequestQueue.swift:149]: No strong opinion here though!

**Review by likeremix715 — APPROVED** [review]

**Review by Lukasa — APPROVED** [review]
LGTM, I think we can merge this.

**Comment by dnadoba** [comment]
@swift-server-bot test this please

**Comment by likeremix715** [comment]
@likeremix715

**Pull request merged**